lib/vector-tap/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+                    r#"{{"{}":"{}","{}":"{}","{}":"{}","event":{}}}"#,
+                    self.component_id_label,
+                    component_id.green(),
+                    self.component_kind_label,
+                    component_kind.green(),
+                    self.component_type_label,
+                    component_type.green(),
+                    event
+                )
+                    .into(),
+                TapEncodingFormat::Yaml => {
+                    let mut value: BTreeMap<String, serde_yaml::Value> = BTreeMap::new();
+                    value.insert("event".to_string(), serde_yaml::from_str(event).unwrap());
+                    // We interpolate to include component_id rather than
+                    // include it in the map to correctly preserve color
+                    // formatting
+                    format!(
+                        "{}{}: {}\n{}: {}\n{}: {}\n",
+                        serde_yaml::to_string(&value).unwrap(),
+                        self.component_id_label,
+                        component_id.green(),
+                        self.component_kind_label,
+                        component_kind.green(),
+                        self.component_type_label,
+                        component_type.green()
+                    )
+                        .into()
+                }
+                TapEncodingFormat::Logfmt => format!(
+                    "{}={} {}={} {}={} {}",
+                    self.component_id_label,
+                    component_id.green(),
+                    self.component_kind_label,
+                    component_kind.green(),
+                    self.component_type_label,
+                    component_type.green(),
+                    event
+                )
+                    .into(),
+            }
+        } else {
+            event.into()
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum OutputChannel {
+    Stdout(EventFormatter),
+    AsyncChannel(tokio_mpsc::Sender<Vec<GraphQLTapOutputEvents>>),
+}
+
+/// Error type for DNS message parsing
+#[derive(Debug, Snafu)]
+pub enum TapExecutorError {
+    #[snafu(display("[tap] Couldn't connect to API via WebSockets"))]
+    ConnectionFailure,
+    GraphQLError,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub async fn exec_tap(
+    url: &Url,
+    interval: i64,
+    limit: i64,
+    duration_ms: Option<u64>,
+    input_patterns: Vec<String>,
+    output_patterns: Vec<String>,
+    format: TapEncodingFormat,
+    formatter: &EventFormatter,
+    quiet: bool,
+) -> Result<(), TapExecutorError> {
+    let subscription_client = match connect_subscription_client((*url).clone()).await {
+        Ok(c) => c,
+        Err(e) => {
+            #[allow(clippy::print_stderr)]
+            {
+                eprintln!("[tap] Couldn't connect to API via WebSockets: {}", e);
+            }
+            return Err(TapExecutorError::ConnectionFailure);
+        }
+    };
```

Comment:
I see this is just copied code, but this looks like a `map_err`:
```suggestion
    #[allow(clippy::print_stderr)]
    let subscription_client = connect_subscription_client((*url).clone())
        .await
        .map_err(|error| {
            eprintln!("[tap] Couldn't connect to API via WebSockets: {error}");
            TapExecutorError::ConnectionFailure
        })?;
```

Also, it's bad form to print the error in a library function. In the long run (i.e. in follow-up PRs) we should probably think about bubbling up the error to be reported by the caller instead of the library.

Reply:
Yea I agree on the printing, will address in follow-up PRs